do_mpc/controller.py

```diff
@@ -30,6 +30,666 @@
+            temp_p = self.P
+            for k in range(self.n_horizon):
+                 K = -np.linalg.inv(np.transpose(B)@temp_p@B+self.R)@np.transpose(B)@temp_p@A
+                 temp_pi = self.Q+np.transpose(A)@temp_p@A-np.transpose(A)@temp_p@B@np.linalg.inv(np.transpose(B)@temp_p@B+self.R)@np.transpose(B)@temp_p@A
+                 temp_p = temp_pi
+            return K
+        
+        #Calculating infinite horizon gain
+        elif self.n_horizon == 0:
+            pi_discrete = solve_discrete_are(A,B, self.Q, self.R)
+            K = -np.linalg.inv(np.transpose(B)@pi_discrete@B+self.R)@np.transpose(B)@pi_discrete@A
+            return K
+    
+    def input_rate_penalization(self):
+        """Computes lqr gain for the input rate penalization mode.
+        
+        This method modifies the state matrix and input matrix according to the input rate penalization method. Due to this objective function also gets modified.
+        The input rate penalization formulation is given as:
+            
+            .. math::
+                x(k+1) = \\tilde{A} x(k) + \\tilde{B}\\Delta u(k)\\\\
+                
+                \\text{where} 
+                \\tilde{A} = \\begin{bmatrix} 
+                                A & B \\\\
+                                0 & I \\end{bmatrix},
+                \\tilde{B} = \\begin{bmatrix} B \\\\
+                             I \\end{bmatrix}
+                            
+        
+        The above formulation is with respect to discrete time system. After formulating the objective, discrete gain is calculated
+        using :py:func:`discrete_gain`
+        
+        :return: Gain matrix :math:`K`
+        :rtype: numpy.ndarray
+        
+
+        """
+        
+        #verifying the given model is daemodel
+        assert self.model.flags['dae2odemodel']==False, 'Already performing input rate penalization without the use of this function'
+        
+        #Verifying input cost matrix for input rate penalization
+        assert self.R_delu.size != 0 , 'set R_delu parameter using set_param() fun.'
+        
+        #Modifying A and B matrix for input rate penalization
+        identity_u = np.identity(np.shape(self.B)[1])
+        zeros_A = np.zeros((np.shape(self.B)[1],np.shape(self.A)[1]))
+        self.A_new = np.block([[self.A,self.B],[zeros_A,identity_u]])
+        self.B_new = np.block([[self.B],[identity_u]])
+        zeros_Q = np.zeros((np.shape(self.Q)[0],np.shape(self.R)[1]))
+        zeros_Ru = np.zeros((np.shape(self.R)[0],np.shape(self.Q)[1]))
+        
+        #Modifying Q and R matrix for input rate penalization
+        self.Q = np.block([[self.Q,zeros_Q],[zeros_Ru,self.R]])
+        self.P = np.block([[self.P,zeros_Q],[zeros_Ru,self.R]])
+        self.R = self.R_delu
+        
+        #Computing gain matrix
+        K = self.discrete_gain(self.A_new, self.B_new)
+        return K
+ 
+    
+    def set_param(self,**kwargs):
```

Comment:
Explain here (and in the general module description) how to obtain infinite horizon LQR. 
I think currently it happens for ``n_horizon = 0`` this is, in my opinion, not very intuitive. 
I would prefer something like ``None`` or even ``np.inf``.